src/formats/jpeg.zig

```diff
@@ -0,0 +1,1084 @@
+        const yDensity = try reader.readIntBig(u16);
+
+        const thumbnailWidth = try reader.readByte();
+        const thumbnailHeight = try reader.readByte();
+
+        if (thumbnailWidth != 0 or thumbnailHeight != 0) {
+            return error.ThumbnailImagesUnsupported;
+        }
+
+        // Make sure there are no application markers after us.
+        if (((try reader.readIntBig(u16)) & 0xFFF0) == @enumToInt(Markers.Application0)) {
+            return error.ExtraneousApplicationMarker;
+        }
+
+        try seek_stream.seekBy(-2);
+
+        return JFIFHeader{
+            .jfifRevision = jfifRevision,
+            .densityUnit = densityUnit,
+            .xDensity = xDensity,
+            .yDensity = yDensity,
+        };
+    }
+};
+
+const QuantizationTable = union(enum) {
+    Q8: [64]u8,
+    Q16: [64]u16,
+
+    pub fn init(precision: u8, reader: ImageReader) !QuantizationTable {
+        // 0 = 8 bits, 1 = 16 bits
+        switch (precision) {
+            0 => {
+                var table = QuantizationTable{ .Q8 = undefined };
+
+                var offset: usize = 0;
+                while (offset < 64) : (offset += 1) {
+                    const value = try reader.readByte();
+                    table.Q8[ZigzagOffsets[offset]] = value;
+                }
+
+                if (JPEG_DEBUG) {
+                    var i: usize = 0;
+                    while (i < 8) : (i += 1) {
+                        var j: usize = 0;
+                        while (j < 8) : (j += 1) {
+                            std.debug.print("{d:4} ", .{table.Q8[i * 8 + j]});
+                        }
+                        std.debug.print("\n", .{});
+                    }
+                }
+
+                return table;
+            },
+            1 => {
+                var table = QuantizationTable{ .Q16 = undefined };
+
+                var offset: usize = 0;
+                while (offset < 64) : (offset += 1) {
+                    const value = try reader.readIntBig(u16);
+                    table.Q16[ZigzagOffsets[offset]] = value;
+                }
+
+                return table;
+            },
+            else => return error.UnknownQuantizationTablePrecision,
+        }
+    }
+};
+
+const HuffmanCode = struct { length_minus_one: u4, code: u16 };
+const HuffmanCodeMap = std.AutoArrayHashMap(HuffmanCode, u8);
+
+const HuffmanTable = struct {
+    allocator: *Allocator,
+
+    code_counts: [16]u8,
+    code_map: HuffmanCodeMap,
+
+    table_class: u8,
+
+    fn generateCodeTable(self: *HuffmanTable, values: [16][]u8) !void {
+        _ = self;
+        var k: usize = 0;
+        var code: usize = 0;
+        var si = values[0].len;
+
+        while (values[k].len != 0) {
+            while (values[k].len == si) {
+                // huffcode[k] = code;
+                code += 1;
+                k += 1;
+            }
+
+            while (values[k].len != si) {
+                code <<= 1;
+                si += 1;
```

Comment:
This block of code does not anything useful right now, does it means to be stored in `self.code_map` ?